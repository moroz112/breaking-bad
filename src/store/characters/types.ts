import { RequestState } from '../types';

export interface ICharacter {
  id: number;
  name: string;
  birthday: string;
  occupation: string[];
  img: string;
  status: string;
  nickname: string;
  appearance: string[];
  portrayed: string;
  category: string[];
}

export interface ICharactersState {
  currentCharacter: ICharacter | null;
  characterByIdGetState: RequestState | null;
  characterByIdGetError: string | null;
}

export const CHARACTER_GET_ONE = 'CHARACTER_GET_ONE';
export const CHARACTER_GET_ONE_SUCCESS = 'CHARACTER_GET_ONE_SUCCESS';
export const CHARACTER_GET_ONE_ERROR = 'CHARACTER_GET_ONE_ERROR';

// Get One Episode
export interface CharacterGetOneAction {
  type: typeof CHARACTER_GET_ONE;
  payload: string;
}

export interface CharacterGetOneSuccessAction {
  type: typeof CHARACTER_GET_ONE_SUCCESS;
  payload: ICharacter;
}

export interface CharacterGetOneErrorAction {
  type: typeof CHARACTER_GET_ONE_ERROR;
  payload: any;
}

export type CharactersActionTypes =
  | CharacterGetOneAction
  | CharacterGetOneSuccessAction
  | CharacterGetOneErrorAction;
