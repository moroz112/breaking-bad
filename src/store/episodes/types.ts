import { RequestState } from '../types';

export enum EpisodeSeries {
  BreakingBad = 'Breaking Bad',
  BetterCallSaul = 'Better Call Saul',
}

export interface IEpisode {
  episode_id: number;
  title: string;
  season: number;
  episode: number;
  air_date: string;
  characters: string[];
  series: EpisodeSeries;
}

export interface IEpisodesState {
  episodes: IEpisode[];
  episodesGetState: RequestState | null;
  episodesGetError: string | null;

  currentEpisode: IEpisode | null;
  episodeByIdGetState: RequestState | null;
  episodeByIdGetError: string | null;
}

export const EPISODES_GET = 'EPISODES_GET';
export const EPISODES_GET_SUCCESS = 'EPISODES_GET_SUCCESS';
export const EPISODES_GET_ERROR = 'EPISODES_GET_ERROR';

export const EPISODE_GET_ONE = 'EPISODE_GET_ONE';
export const EPISODE_GET_ONE_SUCCESS = 'EPISODE_GET_ONE_SUCCESS';
export const EPISODE_GET_ONE_ERROR = 'EPISODE_GET_ONE_ERROR';

export interface EpisodesGetAction {
  type: typeof EPISODES_GET;
  payload: EpisodeSeries;
}

export interface EpisodesGetSuccessAction {
  type: typeof EPISODES_GET_SUCCESS;
  payload: IEpisode[];
}

export interface EpisodesGetErrorAction {
  type: typeof EPISODES_GET_ERROR;
  payload: any;
}

export interface EpisodesGetOneAction {
  type: typeof EPISODE_GET_ONE;
  payload: number;
}

export interface EpisodesGetOneSuccessAction {
  type: typeof EPISODE_GET_ONE_SUCCESS;
  payload: IEpisode;
}

export interface EpisodesGetOneErrorAction {
  type: typeof EPISODE_GET_ONE_ERROR;
  payload: any;
}

export type EpisodesActionTypes =
  | EpisodesGetAction
  | EpisodesGetSuccessAction
  | EpisodesGetErrorAction
  | EpisodesGetOneAction
  | EpisodesGetOneSuccessAction
  | EpisodesGetOneErrorAction;
