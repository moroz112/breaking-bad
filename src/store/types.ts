import { ICharactersState } from './characters/types';
import { IEpisodesState } from './episodes/types';

export enum RequestState {
  Loading,
  Saving,
  Success,
  Error,
}

export interface AppStore {
  episodes: IEpisodesState;
  characters: ICharactersState;
}
