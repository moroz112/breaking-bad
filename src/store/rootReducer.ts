import { combineReducers } from 'redux';
import EpisodesReducer from './episodes/reducer';
import CharactersReducer from './characters/reducer';

const rootReducer = combineReducers({
  episodes: EpisodesReducer,
  characters: CharactersReducer,
});

export default rootReducer;
