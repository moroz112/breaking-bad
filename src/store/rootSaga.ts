import { all, fork } from 'redux-saga/effects';

import episodes from './episodes/sagas';
import characters from './characters/sagas';

// this is main root saga
export default function* root() {
  yield all([
      fork(episodes),
      fork(characters)]
  );
}
