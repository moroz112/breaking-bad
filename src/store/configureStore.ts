import thunk from 'redux-thunk';
import { createStore, applyMiddleware, compose } from 'redux';
import { createLogger } from 'redux-logger';
import createSagaMiddleware from 'redux-saga';

import rootReducer from './rootReducer';
import rootSaga from './rootSaga';

const sagaMiddleware = createSagaMiddleware();

const configureStore = (preloadedState: any) => {
  const enhancers = compose(
    applyMiddleware(thunk, createLogger(), sagaMiddleware),
  );
  const store = createStore(rootReducer, preloadedState, enhancers);
  return store;
};

const store = configureStore({});
sagaMiddleware.run(rootSaga);

export default store;
