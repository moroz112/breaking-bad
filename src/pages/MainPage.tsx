import clsx from 'clsx';
import Button from '../components/Button/Button';
import Table from '../components/Table/Table';
import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import { AppStore, RequestState } from '../store/types';
import { episodesGet } from '../store/episodes/actions';
import { EpisodeSeries, IEpisode } from '../store/episodes/types';

const episodesColumns = [
  {
    key: 'episode',
    dataIndex: 'episode',
    title: 'Episode Number',
  },
  {
    key: 'title',
    dataIndex: 'title',
    title: 'Title',
  },
];

const MainPage = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const { episodes, episodesGetState } = useSelector((store: AppStore) => store.episodes);
  const [series, setSeries] = useState<EpisodeSeries>(EpisodeSeries.BreakingBad);
  const [season, setSeason] = useState(1);
  const [seasonsCount, setSeasonsCount] = useState(1);
  const isLoading = episodesGetState === RequestState.Loading;

  useEffect(() => {
    dispatch(episodesGet(series));
    setSeason(1);
  }, [dispatch, series]);

  useEffect(() => {
    setSeasonsCount(
      episodes?.reduce(
        (acc, currentValue) =>
          Number(currentValue.season) > acc ? Number(currentValue.season) : acc,
        0,
      ) || 1
    );
  }, [episodes]);

  const filteredEpisodes = episodes?.filter((episode) => Number(episode.season) === season)

  const onRowCLick = (entity: IEpisode) => {
    navigate(`/episode/${entity.episode_id}`);
  };

  return (
    <main className="max-w-2xl mx-auto mt-16">
      <div className="flex justify-between mt-8 mb-2">
        <h2 className="text-2xl font-bold leading-7 text-gray-900">Episodes</h2>
        <nav className="flex">
          <div
              className={clsx('cursor-pointer ml-8 px-4', {
                'text-red-600': series === EpisodeSeries.BetterCallSaul,
              })}
              onClick={() => setSeries(EpisodeSeries.BetterCallSaul)}
          >
            Better Call Saul
          </div>
          <div
            className={clsx('cursor-pointer', {
              'text-red-600': series === EpisodeSeries.BreakingBad,
            })}
            onClick={() => setSeries(EpisodeSeries.BreakingBad)}
          >
            Breaking Bad
          </div>
        </nav>
      </div>
      {isLoading ? (
        <h3>Loading...</h3>
      ) : (
        <>
          <div className="flex py-4">
            {Array.from(Array(seasonsCount).keys()).map((index) => (
                <div
                    className={clsx('px-2 cursor-pointer', {'text-red-600': index + 1 === season})}
                    key={index}
                    onClick={() => setSeason(index + 1)}
                >
                  {index + 1}
                </div>
            ))}
          </div>

          {!!episodes.length && (
            <Table
              columns={episodesColumns}
              data={filteredEpisodes}
              keyIndex="episode_id"
              onRowCLick={onRowCLick}
            />
          )}
        </>
      )}
    </main>
  );
};

export default MainPage;
