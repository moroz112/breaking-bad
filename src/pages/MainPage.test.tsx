import React from 'react';
import renderer from 'react-test-renderer';
import { useDispatch, useSelector } from 'react-redux';
import { RequestState } from '../store/types';
import { useParams } from 'react-router-dom';
import MainPage from './MainPage';

jest.mock('react-redux');
jest.mock('react-router');

const mockData = {
  episodesGetState: RequestState.Success,
  episodes: [
    {
      air_date: '03-02-2015',
      characters: [
        'Jimmy McGill',
        'Mike Erhmantraut',
        'Kim Wexler',
        'Howard Hamlin',
        'Chuck McGill',
        'Nacho Varga',
      ],
      episode: '5',
      episode_id: 67,
      season: '1',
      series: 'Better Call Saul',
      title: 'Alpine Shepherd Boy',
    },
  ],
};
beforeEach(() => {
  (useParams as any).mockImplementation(() => ({ id: '1' }));
  (useDispatch as any).mockImplementation(() => () => {});
});

test('MainPage must render loading', () => {
  (useSelector as any).mockImplementation(() => ({ episodesGetState: RequestState.Loading }));
  const component = renderer.create(<MainPage />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('MainPage must render data', async () => {
  (useSelector as any).mockImplementation(() => mockData);
  const component = renderer.create(<MainPage />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
