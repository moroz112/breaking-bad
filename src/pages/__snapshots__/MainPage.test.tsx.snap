// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage must render data 1`] = `
<main
  className="max-w-2xl mx-auto mt-16"
>
  <div
    className="flex justify-between mt-8 mb-2"
  >
    <h2
      className="text-2xl font-bold leading-7 text-gray-900"
    >
      Episodes
    </h2>
    <nav
      className="flex"
    >
      <div
        className="cursor-pointer ml-8 px-4"
        onClick={[Function]}
      >
        Better Call Saul
      </div>
      <div
        className="cursor-pointer text-red-600"
        onClick={[Function]}
      >
        Breaking Bad
      </div>
    </nav>
  </div>
  <div
    className="flex py-4"
  >
    <div
      className="px-2 cursor-pointer text-red-600"
      onClick={[Function]}
    >
      1
    </div>
  </div>
  <div
    className="flex flex-col"
  >
    <div
      className="-my-2 overflow-x-auto sm:-mx-6 lg:-mx-8"
    >
      <div
        className="py-2 align-middle inline-block min-w-full sm:px-6 lg:px-8"
      >
        <div
          className="shadow overflow-hidden border-b border-gray-200 sm:rounded-lg"
        >
          <table
            className="min-w-full divide-y divide-gray-200"
          >
            <thead
              className="bg-gray-50"
            >
              <tr>
                <th
                  className="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
                  scope="col"
                >
                  Episode Number
                </th>
                <th
                  className="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
                  scope="col"
                >
                  Title
                </th>
              </tr>
            </thead>
            <tbody
              className="bg-white divide-y divide-gray-200"
            >
              <tr
                className="cursor-pointer"
                onClick={[Function]}
              >
                <td
                  className="px-6 py-4 whitespace-nowrap text-sm text-gray-500"
                >
                  5
                </td>
                <td
                  className="px-6 py-4 whitespace-nowrap text-sm text-gray-500"
                >
                  Alpine Shepherd Boy
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</main>
`;

exports[`MainPage must render loading 1`] = `
<main
  className="max-w-2xl mx-auto mt-16"
>
  <div
    className="flex justify-between mt-8 mb-2"
  >
    <h2
      className="text-2xl font-bold leading-7 text-gray-900"
    >
      Episodes
    </h2>
    <nav
      className="flex"
    >
      <div
        className="cursor-pointer ml-8 px-4"
        onClick={[Function]}
      >
        Better Call Saul
      </div>
      <div
        className="cursor-pointer text-red-600"
        onClick={[Function]}
      >
        Breaking Bad
      </div>
    </nav>
  </div>
  <h3>
    Loading...
  </h3>
</main>
`;
